From f5645f81364809d3d809c99edddeca055124d163 Mon Sep 17 00:00:00 2001
From: Yuri Konotopov <ykonotopov@gnome.org>
Date: Thu, 14 Dec 2017 20:32:20 +0000
Subject: [PATCH] Mediatek: ignore APs with zero-length SSIDs and zero BSSIDs.

MTK wlan driver may insert such broken APs in search results instead of
debug messages when wifi nvram file does not contains valid device mac.

Change-Id: Ia36c7be3066ab62e7eda220bacea64983b47b472
TODO: this is hack. We should filter scan results somewhere in the driver?
---
 src/drivers/driver_nl80211_scan.c | 3 +++
 wpa_supplicant/bss.c              | 3 +++
 2 files changed, 6 insertions(+)

diff --git a/src/drivers/driver_nl80211_scan.c b/src/drivers/driver_nl80211_scan.c
index c115b6b..fd75540 100644
--- a/src/drivers/driver_nl80211_scan.c
+++ b/src/drivers/driver_nl80211_scan.c
@@ -134,6 +134,9 @@ nl80211_scan_common(struct i802_bss *bss, u8 cmd,
 		if (ssids == NULL)
 			goto fail;
 		for (i = 0; i < params->num_ssids; i++) {
+			if(!params->ssids[i].ssid_len)
+				continue;
+
 			wpa_hexdump_ascii(MSG_MSGDUMP, "nl80211: Scan SSID",
 					  params->ssids[i].ssid,
 					  params->ssids[i].ssid_len);
diff --git a/wpa_supplicant/bss.c b/wpa_supplicant/bss.c
index 3687a2e..a0a3fa3 100644
--- a/wpa_supplicant/bss.c
+++ b/wpa_supplicant/bss.c
@@ -747,6 +747,9 @@ void wpa_bss_update_scan_res(struct wpa_supplicant *wpa_s,
 	if (mesh && mesh[1] <= SSID_MAX_LEN)
 		ssid = mesh;
 
+	if (is_zero_ether_addr(res->bssid))
+		return;
+
 	bss = wpa_bss_get(wpa_s, res->bssid, ssid + 2, ssid[1]);
 	if (bss == NULL)
 		bss = wpa_bss_add(wpa_s, ssid + 2, ssid[1], res, fetch_time);
-- 
2.13.6

